packages/actor-extract-links-extract-tree/lib/FilterNode.ts

```diff
@@ -0,0 +1,194 @@
+      return false;
+    }
+
+    if (node.relation.length === 0) {
+      return false;
+    }
+
+    const query: Algebra.Operation = context.get(KeysInitQuery.query)!;
+    if (!FilterNode.doesNodeExist(query, Algebra.types.FILTER)) {
+      return false;
+    }
+
+    return true;
+  }
+
+  public async run(node: ITreeNode, context: IActionContext): Promise<Map<string, boolean>> {
+    const filterMap: Map<string, boolean> = new Map();
+
+    if (!this.test(node, context)) {
+      return new Map();
+    }
+
+    // Extract the filter expression
+    const filterOperation: Algebra.Expression = (() => {
+      const query: Algebra.Operation = context.get(KeysInitQuery.query)!;
+      return query.input.expression;
+    })();
+
+    // Extract the bgp of the query
+    const queryBody: RDF.Quad[] = FilterNode.findBgp(context.get(KeysInitQuery.query)!);
+    if (queryBody.length === 0) {
+      return new Map();
+    }
+    // Capture the relation from the input
+    const relations: ITreeRelation[] = node.relation!;
+
+    for (const relation of relations) {
+      if (typeof relation.path === 'undefined' || typeof relation.value === 'undefined') {
+        filterMap.set(relation.node, true);
+        continue;
+      }
+      // Find the quad from the bgp that are related to the TREE relation
+      const relevantQuads = FilterNode.findRelavantQuad(queryBody, relation.path.value);
+      if (relevantQuads.length === 0) {
+        filterMap.set(relation.node, true);
+        continue;
+      }
+
+      // Create the binding in relation to the relevant quad
+      const bindings = FilterNode.createBinding(relevantQuads, relation.value.quad);
+      const filterExpression: Algebra.Operation = FilterNode.deleteUnrelevantFilter(filterOperation, bindings);
+      if (filterExpression.args.length === 0) {
+        filterMap.set(relation.node, true);
+        continue;
+      }
+      const evaluator = new AsyncEvaluator(filterExpression);
+      // Evaluate the filter with the relevant quad binding
+      const result: boolean = await evaluator.evaluateAsEBV(bindings);
+      filterMap.set(relation.node, result);
+    }
+    return filterMap;
+  }
+
+  /**
+   *
+   * @param queryBody
+   * @param path
+   * @returns RDF.Quad[]
+   * find the quad that has as predicate a the TREE:path of a relation
+   */
+  private static findRelavantQuad(queryBody: RDF.Quad[], path: string): RDF.Quad[] {
+    const resp: RDF.Quad[] = [];
+    for (const quad of queryBody) {
+      if (quad.predicate.value === path && quad.object.termType === 'Variable') {
+        resp.push(quad);
+      }
+    }
+    return resp;
+  }
+
+  /**
+   * @param filterExpression
+   * @param binding
+   * @returns Algebra.Expression
+   * delete the filters that are not related to TREE relation
+   */
+  private static deleteUnrelevantFilter(filterExpression: Algebra.Expression, binding: Bindings): Algebra.Expression {
+    let newFilterExpression: Algebra.Expression = {
```

Comment:
I don't think so, because I would have to use the method [`createFilter`](https://github.com/joachimvh/SPARQLAlgebra.js/blob/0f4fc8279fa317ef66380846d8f09ff11cf93125/lib/factory.ts#L31) that take as an argument the filter expression.

Reply:
You can use the expression factory methods: https://github.com/joachimvh/SPARQLAlgebra.js/blob/0f4fc8279fa317ef66380846d8f09ff11cf93125/lib/factory.ts#L75-L85